src/platform/intel/ace/platform.c

```diff
@@ -73,6 +77,62 @@ int platform_boot_complete(uint32_t boot_message)
 	return ipc_platform_send_msg(&msg);
 }
 
+/* address where zephyr PM will save memory during D3 transition */
+extern void *global_imr_ram_storage;
+
+/**
+ * @brief Notifier called before every power state transition.
+ * Works on Primary Core only.
+ * @param state Power state being entered.
+ */
+static void notify_pm_state_entry(enum pm_state state)
+{
+	size_t storage_buffer_size;
+
+	if (state == PM_STATE_SOFT_OFF) {
+		/* allocate IMR global_imr_ram_storage */
+		const struct device *tlb_dev = DEVICE_DT_GET(DT_NODELABEL(tlb));
+
+		__ASSERT_NO_MSG(tlb_dev);
+		const struct intel_adsp_tlb_api *tlb_api =
+				(struct intel_adsp_tlb_api *)tlb_dev->api;
+
+		/* get HPSRAM storage buffer size */
+		storage_buffer_size = tlb_api->get_storage_size();
+
+		/* add space for LPSRAM */
+		storage_buffer_size += LP_SRAM_SIZE;
+
+		/* allocate IMR buffer and store it in the global pointer */
+		global_imr_ram_storage = rmalloc(SOF_MEM_ZONE_SYS,
```

Comment:
This is a bit funky to have Zephyr store state into address that is allocated elsewhere. But I guess as long as the L3 heap is implemented on SOF side, we have to do it like this.